orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reindex
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"magma/orc8r/cloud/go/clock"
+	"magma/orc8r/cloud/go/services/state/indexer"
+	"magma/orc8r/cloud/go/services/state/indexer/metrics"
+	state_types "magma/orc8r/cloud/go/services/state/types"
+	merrors "magma/orc8r/lib/go/errors"
+	"magma/orc8r/lib/go/util"
+)
+
+// Major TODOs:
+// TODO(reginawang3495): Add save jobs/indexes # times run
+// TODO(reginawang3495): Remove "Desired" from Indexer.Versions once reindexer_queue and queue are removed
+
+// This Reindexer runs as though it is a singleton
+type reindexerSingleton struct {
+	versioner Versioner
+	store     Store
+	// TODO(reginawang3495) Add some structure to save jobs/indexes # times run
+}
+
+// TODO: to be setup and used in M2 Part C
+const reindexLoopInterval = 5 * time.Second
+
+func NewReindexerSingleton(store Store, versioner Versioner) Reindexer {
+	return &reindexerSingleton{store: store, versioner: versioner}
+}
+
+func (r *reindexerSingleton) Run(ctx context.Context) {
+	const indexerID = ""
+	var sendUpdate func(string) = nil
+
+	batches := r.getReindexBatches(ctx)
+	for {
+		if isCanceled(ctx) {
+			glog.Warning("State reindexing async job canceled")
+			return
+		}
+
+		err := r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+
+		// NOTE: consider sleeping when no err for perf reasons
+		if err != nil {
+			clock.Sleep(failedJobSleep)
+		} else {
+			clock.Sleep(reindexLoopInterval)
+		}
+	}
+}
+
+func (r *reindexerSingleton) RunUnsafe(ctx context.Context, indexerID string, sendUpdate func(string)) error {
+	batches := r.getReindexBatches(ctx)
+	return r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+}
+
+func (r *reindexerSingleton) GetIndexerVersions() ([]*indexer.Versions, error) {
+	return r.versioner.GetIndexerVersions()
+}
```

Comment:
As I understand it, you can just leave this out and it will work implicitly

Reply:
I changed the struct to composition!